Zend/zend_compile.c

```diff
@@ -7463,9 +7656,82 @@ void zend_compile_class_decl(znode *result, zend_ast *ast, zend_bool toplevel) /
 			opline->result.opline_num = -1;
 		}
 	}
+
+	return ce;
 }
 /* }}} */
 
+static void zend_declare_enum_case_constant(zend_ast *ast, zend_class_entry *enum_class)
+{
+	zend_string *enum_case_name = zend_ast_get_str(ast->child[0]);
+	zend_string *enum_class_name = enum_class->name;
+
+	zval class_name_zval;
+	ZVAL_STR_COPY(&class_name_zval, enum_class_name);
+	zend_ast *class_name_ast = zend_ast_create_zval_ex(&class_name_zval, ZEND_NAME_NOT_FQ);
+
+	zval case_name_zval;
+	ZVAL_STR_COPY(&case_name_zval, enum_case_name);
+	zend_ast *case_name_ast = zend_ast_create_zval_ex(&case_name_zval, ZEND_NAME_NOT_FQ);
+
+	zend_ast *case_value_zval_ast = NULL;
+	zend_ast *case_value_ast = ast->child[1];
+	if (case_value_ast != NULL) {
+		znode case_value_node;
+		zend_compile_expr(&case_value_node, case_value_ast);
+		if (case_value_node.op_type != IS_CONST) {
+			zend_error_noreturn(E_ERROR, "Enum case value must be constant");
+		}
+		if (enum_class->enum_primitive_type != Z_TYPE(case_value_node.u.constant)) {
+			zend_error_noreturn(E_ERROR, "Enum case type %s does not match enum type %s", 
+				zend_get_type_by_const(Z_TYPE(case_value_node.u.constant)),
+				zend_get_type_by_const(enum_class->enum_primitive_type));
+		}
+		case_value_zval_ast = zend_ast_create_zval(&case_value_node.u.constant);
+
+		if (enum_class->enum_primitive_type == IS_LONG) {
+			zend_long long_key = Z_LVAL(case_value_node.u.constant);
+			if (zend_hash_index_find(enum_class->enum_case_table, long_key)) {
+				zend_error_noreturn(E_ERROR, "Duplicate enum case value");
+			}
+			zend_hash_index_update(enum_class->enum_case_table, long_key, &case_name_zval);
```

Comment:
```suggestion
			zend_hash_index_add(enum_class->enum_case_table, long_key, &case_name_zval);
```

nit: inconsistent but harmless to use update in one place but add in another

nit: zend_hash_index_add/zend_hash_add returns non-null if the add was successful, which can be used to shorten the check a bit and avoid a second hash lookup, but performance really doesn't matter much here.